include/boost/numeric/ublas/storage.hpp

```diff
@@ -76,7 +76,11 @@ namespace boost { namespace numeric { namespace ublas {
 #pragma warning(pop)
 #endif
                   for (pointer d = data_; d != data_ + size_; ++d)
+#if __cplusplus >= 201703L
```

Comment:
Strictly speaking, this change is needed for C++11 and above (i.e. `__cplusplus >= 201103L`) because there is no guaranteed that a type meeting the Allocator requirements provides a `construct` member.

(The fact nobody noticed this until now suggests nobody tests this code with anything except `std::allocator`, making the allocator support a waste of time.)

Reply:
C++11 standard defines allocator requirements with a `construct` and `destroy` member function, see the [draft version](http://www.open-std.org/jtc1/sc22/wg21/docs/papers/2013/n3690.pdf#subsubsection.17.6.3.5). So I guess it should stay like this.

Reply:
No, you're wrong. Look at the last column of that table. The default shown in the last column means it is **not required** because `allocator_traits` can provide it. Look at the `SimpleAllocator` example on the page after the table of requirements.

Reply:
Jonathan Wakely is right.

"An allocator may constrain the types on which it can be instantiated and the arguments for which its construct member may be called. If a type cannot be used with a particular allocator, the allocator class or the call to construct may fail to instantiate."

These were the quotes in the draft with the example below.

Reply:
That quote isn't actually relevant here, that's talking about something else. The relevant wording is in paragraph 2, before the table, where it says "A requirement is optional if the last column of Table 28 specifies a default for a given expression."

Since `construct` and `destroy` have defaults in the last column, they are optional.

Reply:
yep, `construct` and `destroy` are optional as they have a specified default.

Reply:
Thanks for the information I didn't know about it earlier as I used to create my own allocator similar to the given example and use `std::allocator_traits` instead or I rarely used it.

Reply:
I've created #101 for this problem, since it predates #96 and affects all of C++11 and later, not just C++20.

Reply:
N.B. you can also refer to https://en.cppreference.com/w/cpp/named_req/Allocator which makes it clear which parts of the requirements are optional.

Reply:
Thanks for clarification and for creating the issue https://github.com/boostorg/ublas/issues/101.

Reply:
Changed the hotfix accordingly.